torchvision/csrc/io/decoder/gpu/demuxer.h

```diff
@@ -0,0 +1,205 @@
+    timeBase = av_q2d(rTimeBase);
+    userTimeScale = timeScale;
+
+    bMp4H264 = eVideoCodec == AV_CODEC_ID_H264 &&
+        (!strcmp(fmtCtx->iformat->long_name, "QuickTime / MOV") ||
+         !strcmp(fmtCtx->iformat->long_name, "FLV (Flash Video)") ||
+         !strcmp(fmtCtx->iformat->long_name, "Matroska / WebM"));
+    bMp4HEVC = eVideoCodec == AV_CODEC_ID_HEVC &&
+        (!strcmp(fmtCtx->iformat->long_name, "QuickTime / MOV") ||
+         !strcmp(fmtCtx->iformat->long_name, "FLV (Flash Video)") ||
+         !strcmp(fmtCtx->iformat->long_name, "Matroska / WebM"));
+    bMp4MPEG4 = eVideoCodec == AV_CODEC_ID_MPEG4 &&
+        (!strcmp(fmtCtx->iformat->long_name, "QuickTime / MOV") ||
+         !strcmp(fmtCtx->iformat->long_name, "FLV (Flash Video)") ||
+         !strcmp(fmtCtx->iformat->long_name, "Matroska / WebM"));
+
+    av_init_packet(&pkt);
+    pkt.data = NULL;
+    pkt.size = 0;
+    av_init_packet(&pktFiltered);
+    pktFiltered.data = NULL;
+    pktFiltered.size = 0;
+
+    if (bMp4H264) {
+      const AVBitStreamFilter* bsf = av_bsf_get_by_name("h264_mp4toannexb");
+      if (!bsf) {
+        printf("FFmpeg error: %d, av_bsf_get_by_name() failed\n", __LINE__);
+        return;
+      }
+      check_for_errors(av_bsf_alloc(bsf, &bsfCtx));
+      avcodec_parameters_copy(
+          bsfCtx->par_in, fmtCtx->streams[iVideoStream]->codecpar);
+      check_for_errors(av_bsf_init(bsfCtx));
+    }
+    if (bMp4HEVC) {
+      const AVBitStreamFilter* bsf = av_bsf_get_by_name("hevc_mp4toannexb");
+      if (!bsf) {
+        printf("FFmpeg error: %d, av_bsf_get_by_name() failed\n", __LINE__);
+        return;
+      }
+      check_for_errors(av_bsf_alloc(bsf, &bsfCtx));
+      avcodec_parameters_copy(
+          bsfCtx->par_in, fmtCtx->streams[iVideoStream]->codecpar);
+      check_for_errors(av_bsf_init(bsfCtx));
+    }
+  }
+  ~Demuxer() {
+    if (!fmtCtx) {
+      return;
+    }
+    if (pkt.data) {
+      av_packet_unref(&pkt);
+    }
+    if (pktFiltered.data) {
+      av_packet_unref(&pktFiltered);
+    }
+    if (bsfCtx) {
+      av_bsf_free(&bsfCtx);
+    }
+    avformat_close_input(&fmtCtx);
+    if (dataWithHeader) {
+      av_free(dataWithHeader);
+    }
+  }
+
+  AVCodecID get_video_codec() {
+    return eVideoCodec;
+  }
+
+  bool demux(uint8_t** video, unsigned long* videoBytes) {
+    if (!fmtCtx) {
+      return false;
+    }
+    *videoBytes = 0;
+
+    if (pkt.data) {
+      av_packet_unref(&pkt);
+    }
+    int e = 0;
+    while ((e = av_read_frame(fmtCtx, &pkt)) >= 0 &&
+           pkt.stream_index != iVideoStream) {
+      av_packet_unref(&pkt);
+    }
+    if (e < 0) {
+      return false;
+    }
+
+    if (bMp4H264 || bMp4HEVC) {
+      if (pktFiltered.data) {
+        av_packet_unref(&pktFiltered);
+      }
+      check_for_errors(av_bsf_send_packet(bsfCtx, &pkt));
+      check_for_errors(av_bsf_receive_packet(bsfCtx, &pktFiltered));
+      *video = pktFiltered.data;
+      *videoBytes = pktFiltered.size;
```

Comment:
Let's double-check that we don't have leaks in here. I think it should be fine, but just checking

Reply:
I ran cuda-memcheck and didn't get any memory leaks.

Reply:
This code runs on the CPU. I might be wrong but I though that cuda memcheck was for checking CUDA memory?

Reply:
You are right, I'll run valgrind to make sure there is no CPU memory leak.